aws/resource_aws_security_group.go

```diff
@@ -1143,6 +1143,118 @@ func matchRules(rType string, local []interface{}, remote []map[string]interface
 	return saves
 }
 
+func collapseRules(ruleset string, rules []interface{}) []interface{} {
```

Comment:
These new functions are quite complicated to understand with quick reading. Can we please:
* Add [Go documentation](https://blog.golang.org/godoc-documenting-go-code) explaining why each function  exists and what they are intended to accomplish
* Add unit testing to explain the usage a little better and prevent future regressions

One other minor nitpick: the provider currently uses a very flat Go package structure, which means function names span all AWS services -- could we rename these new functions to be more specific as other services also have "rules", e.g. `collapaseEc2SecurityGroupRules`